modin/core/dataframe/algebra/groupby.py

```diff
@@ -373,16 +412,196 @@ def try_filter_dict(agg_func, df):
+            agg_dict
+        ):
+            # Filter dictionary
+            dict_to_add = (
+                external_aggs if cls.is_external_aggregation(func) else native_aggs
+            )
+
+            new_value = func if func_name is None else (func_name, func)
+            old_value = dict_to_add.get(col)
+
+            if old_value is not None:
+                assert isinstance(
+                    old_value, list
+                ), f"`{col_renaming_required=}` must be `True` when multiple aggregations per columns is specified."
+                old_value.append(new_value)
+            else:
+                dict_to_add[col] = [new_value] if col_renaming_required else new_value
+
+            # Construct resulting columns
+            if col_renaming_required:
+                func_name = str(func) if func_name is None else func_name
+                result_columns.append(
+                    (*(col if isinstance(col, tuple) else (col,)), func_name)
+                )
+            else:
+                result_columns.append(col)
+
+        result_columns = pandas.Index(result_columns)
+
+        def aggregate_on_dict(grp_obj, *args, **kwargs):
+            """Aggregate the passed groupby object."""
+            if len(native_aggs) == 0:
+                native_agg_res = None
+            elif grp_has_id_level:
+                # Adding the 'id' level to the aggregation keys so they match `grp_obj` columns
+                native_aggs_modified = {
+                    (cls.ID_LEVEL_NAME, *key): value
+                    for key, value in native_aggs.items()
+                }
+                native_agg_res = grp_obj.agg(native_aggs_modified)
+                # Dropping the 'id' level from the resulted frame
+                native_agg_res.columns = native_agg_res.columns.droplevel(0)
+            else:
+                native_agg_res = grp_obj.agg(native_aggs)
+
+            external_results = []
+            insert_id_levels = False
+
+            for col, func, func_name, col_renaming_required in walk_aggregation_dict(
+                external_aggs
+            ):
+                if grp_has_id_level:
+                    cols_without_ids = grp_obj.obj.columns.droplevel(0)
+                    col_pos = (
+                        # We may have multiple columns matching the `col` in
+                        # a MultiIndex case, that's why use `.get_locs` here
+                        cols_without_ids.get_locs(col)
+                        if isinstance(cols_without_ids, pandas.MultiIndex)
+                        # `pandas.Index` doesn't have `.get_locs` method
+                        else cols_without_ids.get_loc(col)
+                    )
+                    agg_key = grp_obj.obj.columns[col_pos]
+                else:
+                    agg_key = [col]
+
+                result = func(grp_obj[agg_key])
+                result_has_id_level = result.columns.names[0] == cls.ID_LEVEL_NAME
```

Comment:
Why do we need to check that? Shouldn't `result_has_id_level` always be equal to `grp_has_id_level`?

Reply:
Added an in-code comment to make it clearer:
```python
result = func(grp_obj[agg_key])                
# The `func` may have discarded an ID-level if there were any.
# So checking for this again.
result_has_id_level = result.columns.names[0] == cls.ID_LEVEL_NAME
```